### PR TITLE
Bitstream Tools: Add Spartan6 support

### PR DESCRIPTION
**Pull request by tmichalak**
This PR supersedes #1025 and depends on #1091  
Spartan6 support has been added with the use of a more generic bitstream tools API.
Created a Spartan6 version of the unit tests that have been adapted to the new API.

### CODE REVIEW

**Comment by litghost** [comment]
CI failure looks like a formatting issue.

**Comment by tmichalak** [comment]
@litghost 
I would like to merge #1090 before this one. Adding Spartan6 and then other architectures won't need explicit template instantiations in the source files anymore.

**Comment by litghost** [comment]
> @litghost
> I would like to merge #1090 before this one. Adding Spartan6 and then other architectures won't need explicit template instantiations in the source files anymore.
> Of course I will need to rebase on the future new master.

#1090 is merged, let's rebase this PR and get it green on CI.

**Comment by litghost** [comment]
Let's rebase this one again.  Can any of the changes in this PR take notes from the changes made to the US or US+ changes?

**Comment by tmichalak** [comment]
Ha, you beat me with the comment about rebasing by 9 seconds:)

File: tools/bittool.cc [hunk 1]
```diff
@@ -170,14 +170,28 @@ struct DeviceIdGetter {
 			            ArchType::ConfRegType::IDCODE);
 		    });
 		if (idcode_packet != reader->end()) {
-			if (idcode_packet->data().size() != 1) {
-				std::cerr
-				    << "Write to IDCODE with word_count != 1"
-				    << std::endl;
-				return 1;
+			if (std::is_same<ArchType, xilinx::Spartan6>::value) {
```
Comment (litghost) [inline tools/bittool.cc]: This seems weird.  Can think of a way to make a better way to do this in bittool and bitread (which both have a similar block)?

Comment (tmichalak) [reply tools/bittool.cc]: At the time when this was done, I tried to leave the most of the untouched and focused on the relevant part. 
However, this part, and a probably also some other of the original code could use some refactor.

**Comment by tmichalak** [comment]
> Can any of the changes in this PR take notes from the changes made to the US or US+ changes?
I don't think there is much more we can gain with further refactor, at least with the current software architecture. The differences between Spartan6 and Series7, US and US+ are much bigger than the rest among themselves.

File: lib/include/prjxray/xilinx/frames.h [hunk 2]
```diff
@@ -67,13 +66,18 @@ int Frames<ArchType>::readFrames(const std::string& frm_file_str) {
...
 		    absl::StrSplit(frame_delta.second, ',');
 
-		if (frame_data_strings.size() != ArchType::words_per_frame) {
-			std::cerr << "Frame " << std::hex << frame_address
-			          << ": found " << std::dec
-			          << frame_data_strings.size()
-			          << " words instead of "
-			          << ArchType::words_per_frame << std::endl;
-			continue;
+		// Spartan6's IOB frames can have different word count
```
Comment (litghost) [inline lib/include/prjxray/xilinx/frames.h]: Should each arch have a function that checks this?  On xc7, US and US+, it will be a straight forward check, and spartan6 will have a data lookup?

File: lib/include/prjxray/xilinx/spartan6/block_type.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+#include <ostream>
+
+#include <yaml-cpp/yaml.h>
+
+namespace prjxray {
+namespace xilinx {
+namespace spartan6 {
+
+enum class BlockType : unsigned int {
```
Comment (litghost) [inline lib/include/prjxray/xilinx/spartan6/block_type.h]: Comment where this comes from

File: lib/include/prjxray/xilinx/spartan6/command.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#ifndef PRJXRAY_LIB_XILINX_SPARTAN6_COMMAND_H_
+#define PRJXRAY_LIB_XILINX_SPARTAN6_COMMAND_H_
+
+namespace prjxray {
+namespace xilinx {
+namespace spartan6 {
+
+enum class Command : uint32_t {
```
Comment (litghost) [inline lib/include/prjxray/xilinx/spartan6/command.h]: Comment where this comes from

File: lib/xilinx/spartan6/frame_address.cc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace prjxray {
+namespace xilinx {
+namespace spartan6 {
+
+FrameAddress::FrameAddress(spartan6::BlockType block_type,
+                           uint8_t row,
+                           uint8_t column,
+                           uint16_t minor) {
+	address_ = bit_field_set(0, 31, 28, block_type);
```
Comment (litghost) [inline lib/xilinx/spartan6/frame_address.cc]: Comment where this comes from

File: lib/include/prjxray/xilinx/frames.h (hunk 2 above)
Comment (tmichalak) [reply lib/include/prjxray/xilinx/frames.h]: What kind of check do you have in mind for Spartan6? Do you mean something like checking if the current frames is indeed an IOB frame and if not perform the same check as for all other architectures?

Comment (litghost) [reply lib/include/prjxray/xilinx/frames.h]: Sort of.  Is the IOB frame fully variable, or just a different number?

Comment (litghost) [reply lib/include/prjxray/xilinx/frames.h]: Let me rephrase.  Is there a way to write this check that is valid for all spartan 6 frame addresses?  Why not implement that check?

Comment (tmichalak) [reply lib/include/prjxray/xilinx/frames.h]: Unsure. I just read that the IOB frame length can differ, but it didn't say exactly how and when. In my particular case it had something around 20 when the regular frames is 65. But I can't tell what the number is gonna be for other parts. So at this moment I can't really think of any generic check...

Comment (litghost) [reply lib/include/prjxray/xilinx/frames.h]: Ok, let's file an issue to investigate and fix this later.

Comment (tmichalak) [reply lib/include/prjxray/xilinx/frames.h]: Filed #1113.
